### PR TITLE
add physics component

### PR DESCRIPTION
**Pull request by minggo**
The corresponding issue is #13404.

### CODE REVIEW

**Comment by minggo** [comment]
@WenhaiLin 
Could you please review it too?

**Comment by WenhaiLin** [comment]
@minggo Please fix conflicts.

**Comment by minggo** [comment]
Ok, i fixed the conflict.

**Comment by minggo** [comment]
[ci rebuild]

**Comment by minggo** [comment]
@WenhaiLin 
Do you have any comments?
I will merge it if not.

**Comment by minggo** [comment]
> I suggest to use C++ 11 smart pointer rather than the raw delete operators.

What did you mean?

> Does every scene should own a PhysicsManager or just a PhysicWorld?

PhyiscsWorld is managed by PhysicsManager.

**Comment by zilongshanren** [comment]
@minggo  
I mean prefer unique_ptr to raw pointer new & delete.

**Comment by minggo** [comment]
@zilongshanren 
The pointer is used internally, so there is not need to use smart pointer.

**Comment by minggo** [comment]
I will send another squashed PR after it passes compiling.

File: cocos/physics/CCComponentPhysics2d.cpp [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        ret->autorelease();
+    
+    return ret;
+}
+
+ComponentPhysics2d::ComponentPhysics2d()
+: _physicsBody(nullptr)
+, _physicsRotation(0.0f)
+, _ownerOriginRotation(0.0f)
+{}
```
Comment (WenhaiLin) [inline cocos/physics/CCComponentPhysics2d.cpp]: You should initialize the name of component.

File: cocos/physics/CCComponentPhysics2d.cpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        _physicsBody->_componentBelongsTo = nullptr;
+        _physicsBody->release();
+        _physicsBody = nullptr;
+    }
+}
+
+void ComponentPhysics2d::addToPhysicsManager()
+{
+    if (_owner && _owner->getScene())
+        _owner->getScene()->getPhysicsManager()->addPhysicsComponent(this);
```
Comment (WenhaiLin) [inline cocos/physics/CCComponentPhysics2d.cpp]: Please avoid calling a costly method(`getScene()`) multiple times.

File: cocos/physics/CCComponentPhysics2d.cpp (hunk 1 above)
Comment (minggo) [inline cocos/physics/CCComponentPhysics2d.cpp]: Yep, good catch.

File: cocos/physics/CCComponentPhysics2d.cpp (hunk 2 above)
Comment (minggo) [inline cocos/physics/CCComponentPhysics2d.cpp]: Ok.

**Comment by minggo** [comment]
I sent another squashed PR in #13709.

**Pull request closed without merging**